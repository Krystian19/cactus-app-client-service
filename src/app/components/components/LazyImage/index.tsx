import React from 'react';
import LoadingSpinner from '../LoadingSpinner';

type PropType = {
  src: String,
  errorSrc: String,
  className: String,
  alt: String,
  noLoadingSpinner: Boolean,
}

type StateType = {
  isMounted: Boolean,
  isLoaded: Boolean,
  loadError: Boolean
}

export default class LazyImage extends React.Component<PropType, StateType> {
  constructor(props) {
    super(props);

    this.state = {
      isMounted: false,
      isLoaded: false,
      loadError: false,
    };
  }

  public static defaultProps = {
    noLoadingSpinner: false,
  };

  componentDidMount() {
    const { src } = this.props;

    // Set component state to mounted
    this.setState({ isMounted: true });

    // Force browser to load the requested image
    // and call onload event when/if done
    const img = new Image();
    img.onload = this.imageDidLoad;
    img.onerror = this.imageDidHaveLoadError;
    img.src = String(src);
  }

  imageDidLoad = () => {
    this.setState({ isLoaded: true });
  }

  imageDidHaveLoadError = () => {
    this.setState({ loadError: true });
  }

  render() {
    const {
      src,
      errorSrc,
      className,
      alt,
      noLoadingSpinner,
    } = this.props;

    const { isMounted, isLoaded, loadError } = this.state;

    // If image encountered an error during the load process
    // return the designated error image.
    if (loadError) {
      return (
        <img
          src={String(errorSrc)}
          alt={String(alt)}
          className={String(className)}
        />
      );
    }

    if (
      (!isMounted) // If element is not mounted yet
      || (!isLoaded) // Or if image has not loaded yet
    ) {
      // If no spinner animation is desired
      if (noLoadingSpinner) return (null);

      // Otherwise return the spinner
      return <LoadingSpinner />;
    }

    // Return proper image when is not mounted
    return (
      <img
        src={String(src)}
        alt={String(alt)}
        className={String(className)}
      />
    );
  }
}
