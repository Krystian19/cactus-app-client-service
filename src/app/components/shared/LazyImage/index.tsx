import React from 'react';
import base64Content from '../../../utils/base64Content';

type PropType = {
  src: string;
  errorSrc: string;
  className: string;
  alt: string;
  noLoadingPlaceholder: boolean;
  posterPlaceholder: boolean;
  customLoadingPlaceholderSrc?: string;
};

type StateType = {
  isMounted: boolean;
  isLoaded: boolean;
  loadError: boolean;
};

export default class LazyImage extends React.Component<PropType, StateType> {
  state = {
    isMounted: false,
    isLoaded: false,
    loadError: false,
  };

  public static defaultProps = {
    noLoadingPlaceholder: false,
    posterPlaceholder: false,
  };

  componentDidMount() {
    const { src } = this.props;

    // Set component state to mounted
    this.setState({ isMounted: true });

    // Force browser to load the requested image
    // and call onload event when/if done
    const img = new Image();
    img.onload = this.imageDidLoad;
    img.onerror = this.imageDidHaveLoadError;
    img.src = String(src);
  }

  imageDidLoad = () => {
    this.setState({ isLoaded: true });
  }

  imageDidHaveLoadError = () => {
    this.setState({ loadError: true });
  }

  render() {
    const {
      src,
      errorSrc,
      className,
      alt,
      noLoadingPlaceholder,
      posterPlaceholder,
      customLoadingPlaceholderSrc,
    } = this.props;

    const { isMounted, isLoaded, loadError } = this.state;

    // If image encountered an error during the load process
    // return the designated error image.
    if (loadError) {
      return (
        <img
          src={String(errorSrc)}
          alt={String(alt)}
          className={String(className)}
        />
      );
    }

    if (
      (!isMounted) // If element is not mounted yet
      || (!isLoaded) // Or if image has not loaded yet
    ) {
      // If no spinner animation is desired
      if (noLoadingPlaceholder) return (null);

      // If a custom placeholder source is provided
      if (customLoadingPlaceholderSrc) {
        return (
          <img
            src={String(customLoadingPlaceholderSrc)}
            alt={String(alt)}
            className={String(className)}
          />
        );
      }

      // If a poster placeholder is desired
      if (posterPlaceholder) {
        return (
          <img
            src={base64Content.cactus_poster_placeholder}
            alt={String(alt)}
            className={String(className)}
          />
        );
      }

      // Otherwise return the thumbnail placeholder
      return (
        <img
          src={String(base64Content.cactus_thumbnail_placeholder)}
          alt={String(alt)}
          className={String(className)}
        />
      );
    }

    // Return proper image when is not mounted
    return (
      <img
        src={String(src)}
        alt={String(alt)}
        className={String(className)}
      />
    );
  }
}
