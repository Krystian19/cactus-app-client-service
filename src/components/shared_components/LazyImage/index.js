import React, { Component } from 'react';
import LoadingSpinner from '../LoadingSpinner';

export default class LazyImage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isMounted: false,
      isLoaded: false,
    };
  }

  componentDidMount() {
    const { src } = this.props;

    // Set component state to mounted
    this.setState({ isMounted: true });

    // Force browser to load the requested image
    // and call onload event when/if done
    const img = new Image();
    img.src = src;
    img.onload = this.imageDidLoad;
  }

  imageDidLoad = () => {
    this.setState({ isLoaded: true });
  }

  render() {
    const {
      src,
      errorSrc,
      className,
      alt,
      noLoadingSpinner,
    } = this.props;
    const { isMounted, isLoaded } = this.state;

    // If element is not mounted yet or
    // image has not loaded yet
    if ((!isMounted) || (!isLoaded)) {
      // If no spinner animation is desired
      if (noLoadingSpinner) return (null);

      // Otherwise return the spinner
      return <LoadingSpinner />;
    }

    // TODO: Handle load errors

    // Return proper image when is not mounted
    return (
      <img
        src={src}
        alt={alt}
        className={className}
      />
    );
  }
}
