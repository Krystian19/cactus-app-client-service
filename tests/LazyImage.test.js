import React from 'react';
import renderer from 'react-test-renderer';
import LazyImage from '../src/components/shared_components/LazyImage';

const { test, expect } = window;

test('Tests are working well', () => {
  const component = renderer.create(
    <LazyImage
      src="/img_cdn/test.jpg"
      errorSrc="/img_cdn/test.jpg"
      alt="thumbnail"
      className="anime-small-thumbnail fade-in"
      noLoadingSpinner
    />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
