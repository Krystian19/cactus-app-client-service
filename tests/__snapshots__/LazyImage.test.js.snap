// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImage component is rendering properly 1`] = `null`;

exports[`Tests are working well 1`] = `null`;
